Sources/NIO/FileHandle.swift

```diff
@@ -84,11 +84,20 @@ extension NIOFileHandle {
     /// Open a new `NIOFileHandle`.
     ///
     /// - parameters:
-    ///     - path: the path of the file to open. The ownership of the file descriptor is transferred to this `NIOFileHandle` and so it will be closed once `close` is called.
+    ///     - path: the path of the file to open for reading. The ownership of the file descriptor is transferred to this `NIOFileHandle` and so it will be closed once `close` is called.
     public convenience init(path: String) throws {
         let fd = try Posix.open(file: path, oFlag: O_RDONLY | O_CLOEXEC)
         self.init(descriptor: fd)
     }
+
+    /// Open a new `NIOFileHandle`.
+    ///
+    /// - parameters:
+    ///     - forWritingAtPath: the path of the file to open for writing. The ownership of the file descriptor is transferred to this `NIOFileHandle` and so it will be closed once `close` is called. If file does not exist, it will be created with default permissions 0644.
+    public convenience init(forWritingAtPath path: String, mode: UInt16 = S_IWUSR | S_IRUSR | S_IRGRP | S_IROTH) throws {
```

Comment:
could we have a nicer mode argument here? The POSIX API isn't that nice I think. Also the `forWritingAtPath` looks a bit too much like old day cocoa ;).

Also that kind of mode only matters when _creating_ a new file.

What about an API like this:

```swift

    enum Mode {
        case .readOnly
        case .writeOnly
        case .readWrite
    } // or as OptionSet?
    public struct Flags {
        private posixMode: UInt16

        public static default: Flags = .init(posixMode: 0)

        public func allowFileCreation(posixMode: UInt16) {
            return .init(posixMode: posixFlags)
        }
    }
    public init(path: String, mode: Mode = .read, flags: Flags = .default) { ...}
```

that way first of all we don't need a _new_ API, but we can extend the existing API. Also we don't make the weird POSIX mode the only game in town and also the POSIX mode is only required when it's actually used (on file creation, if that's allowed).

Reply:
good idea, one question: do you propose to expose POSIX flags as something more readable (`.readOnly == O_RDONLY`) or introduce our own modes that map to sensible set of POSIX flags (`.reading = O_RDONLY | O_CLOEXEC`)?

Reply:
`forWritingAtPath` looks exactly like old day cocoa :D

Reply:
@artemredkin I'd expose some options as more readable things. Like `.readOnly` should be `O_RDONLY` but `O_CLOEXEC` we should always set, there's no point in not setting `O_CLOEXEC`.

Reply:
Does this look better?